src/launchpads/MK3/LaunchpadMK3.ts

```diff
@@ -0,0 +1,242 @@
+
+    this.output.sendMessage(sysExMessage);
+  }
+
+  /**
+   * @inheritDoc
+   */
+  setButtonColor(button: number|number[], color: number[]): void {
+    if (!Array.isArray(color) || color.length !== 3) {
+      throw new Error('Invalid color settings supplied');
+    }
+
+    // make sure the launchpad understands the colors we provide
+    if (color.some(value => value > 127 || value < 0)) {
+      throw new Error('RGB color is invalid, please make sure the color values are in range 0-127 (Hint: you can use colors.colorFromRGB as a helper for that');
+    }
+
+    const [r, g, b] = color;
+    const buttonMapped = this.mapButtonFromXy(button);
+
+    this.sendSysEx(...SysEx.setButtonColors(ButtonColor.rgb(
+      buttonMapped, r, g, b)));
+  }
+
+  /**
+   * @inheritDoc
+   */
+  flash(button: number|number[], color: number, colorB: number = 0): void {
+    const buttonMapped = this.mapButtonFromXy(button);
+
+    this.sendSysEx(...SysEx.setButtonColors(ButtonColor.flash(buttonMapped, color, colorB)));
+  }
+
+  /**
+   * @inheritDoc
+   */
+  pulse(button: number|number[], color: number): void {
+    const buttonMapped = this.mapButtonFromXy(button);
+
+    this.sendSysEx(...SysEx.setButtonColors(ButtonColor.pulse(buttonMapped, color)));
+  }
+
+  /**
+   * @inheritDoc
+   */
+  allOff(): void {
+    this.sendSysEx(14, 0);
+  }
+
+  /**
+   * @inheritDoc
+   */
+  closePorts(): void {
+    this.logDebug('Closing ports');
+
+    this.allOff();
+    this.input.closePort();
+    this.output.closePort();
+  }
+
+  /**
+   * @inheritDoc
+   */
+  eventNames(): string[] {
+    return [
+      'ready',
+      'rawMessage',
+      'buttonDown',
+      'buttonUp',
+    ];
+  }
+
+  /**
+   * @inheritDoc
+   */
+  parseButtonToXy(state: number, note: number): number[]|number {
+    if (!this.options.xyMode) {
+      return note;
+    }
+
+    const row = Math.floor(note / 10);
+    const col = (note - 1) % 10;
+
+    return [col, 9 - row];
+  }
+
+  /**
+   * @inheritDoc
+   */
+  mapButtonFromXy(xy: number[]|number): number {
+    const [x, y] = Array.isArray(xy) ? xy : [xy, 0];
+
+    if (!this.options.xyMode) {
+      return x;
+    }
+
+    return (9 - y) * 10 + x + 1;
+  }
+
+  private setupMessageHandler(): void {
+    this.input.on('message', (deltaTime: number, message: number[]) => {
+      this.logDebug(`m: ${message} d: ${deltaTime}`);
+      this.internalMessageHandler(message);
+    });
+  }
+
+  private internalMessageHandler(message: number[]): void {
+    this.emit('rawMessage', message);
+
+    const [state, note, value] = message;
+    const button = this.parseButtonToXy(state, note);
+
+    const upDown = Boolean(value) ? 'Down' : 'Up';
+    this.emit(`button${upDown}`, button, value);
+  }
+
+  private logDebug(...message: any[]): void {
+    if (this.options.debug) {
+      console.log('[Launchpad Debug]', ...message);
+    }
+  }
+}
+
+/**
+ * Specify the color of a single button
+ */
+export class ButtonColor {
```

Comment:
Could you put this in its own file? MIght be nice to have an abstract class for all launchpads

Reply:
I've been nerd-sniped by this and would love to do a more generic interface in general.

How about something like the following?

```ts
export type Button = [number, number]; // (x, y)

export type Color = 
    | { color: number }
    | { flash: number, flashFrom?: number }  // default 0
    | { pulse: number }
    | { r: number, g: number, b: number } // range [0..1]
    ;

class GenericLaunchPad {
    public setButtonColor(button: Button, color: Color | number | [number, number, number]);
    public setAllButtons(colors: Color[]); // Array of length 81
}
```

Reply:
Genericizing SysEx messages is harder, I see.

For the MK3, the protocol is (symbolically)

```
setButtons([
  [button, palette|flash|pulse|rgb],
  [button, palette|flash|pulse|rgb],
  [button, palette|flash|pulse|rgb],
  ...
)
```

For the MK2, the protocol is:

```
setButtonsPalette([button, palette], [button, palette], ...)
setButtonsFlash([button, flash], [button, flash], ...)
setButtonsPulse([button, pulse], [button, pulse], ...)
setButtonsRGB([button, r, g, b], [button, r, g, b], ...)
```

So it seems not just the message numbers, but even what they can represent is version-specific.

Reply:
> I've been nerd-sniped by this and would love to do a more generic interface in general.
> 
> How about something like the following?
> 
> ```ts
> export type Button = [number, number]; // (x, y)
> 
> export type Color = 
>     | { color: number }
>     | { flash: number, flashFrom?: number }  // default 0
>     | { pulse: number }
>     | { r: number, g: number, b: number } // range [0..1]
>     ;
> 
> class GenericLaunchPad {
>     public setButtonColor(button: Button, color: Color | number | [number, number, number]);
>     public setAllButtons(colors: Color[]); // Array of length 81
> }
> ```

I like this

Reply:
> Genericizing SysEx messages is harder, I see.
> 
> For the MK3, the protocol is (symbolically)
> 
> ```
> setButtons([
>   [button, palette|flash|pulse|rgb],
>   [button, palette|flash|pulse|rgb],
>   [button, palette|flash|pulse|rgb],
>   ...
> )
> ```
> 
> For the MK2, the protocol is:
> 
> ```
> setButtonsPalette([button, palette], [button, palette], ...)
> setButtonsFlash([button, flash], [button, flash], ...)
> setButtonsPulse([button, pulse], [button, pulse], ...)
> setButtonsRGB([button, r, g, b], [button, r, g, b], ...)
> ```
> 
> So it seems not just the message numbers, but even what they can represent is version-specific.

That's a shame, just leave it if it's impossible